src/Hooks.ts

```diff
@@ -143,3 +145,75 @@ export function useFocusState() {
 
   return focusState;
 }
+
+type EffectCallback = (() => void) | (() => () => void);
+
+// Inspired by same hook from react-navigation v5
+// See https://github.com/react-navigation/hooks/issues/39#issuecomment-534694135
+export const useFocusEffect = (callback: EffectCallback) => {
+  const navigation = useNavigation();
+  const getCallback = useGetter(callback);
```

Comment:
I think using a getter like this works for simple event listeners where you just need to execute a function on an event, but for this case, it's not correct.

For example:

```js
useFocusEffect(
  React.useCallback(() => API.subscribeToLikes(messageId, setLikeCount), [
    messageId,
  ])
);
```

Here, we need to re-subscribe to the API when:
- screen becomes focused
- `messageId` changes

But since the effect doesn't re-run when the callback changes here, it's going to stay subscribed to old `messageId` even if it has changed until the screen blurs and focuses again. IMO it needs to be changed to pass this callback in the dep array to be correct.

Reply:
hmmm, not totally sure to understand what you mean here. 

The point is to avoid unnecessary unsub/resub, but only subscribe once, and let the listener fire the latest callback. 

We don't need to unsub/resub on messageId change because if the closure changes, the listener will execute the latest one, which has captured the latest messageId.

I'm using this pattern in many places to handle stale closure issues and so far it worked great.

Reply:
> We don't need to unsub/resub on messageId change because if the closure changes, the listener will execute the latest one, which has captured the latest messageId.

The function you're referring to in "if the closure changes" is: `() => API.subscribeToLikes(messageId, setLikeCount)`

If it changes, the listener will execute the latest one. Yes. But it means, the next time this function will get executed is when a `willFocus` event fires, not when `messageId` changes. Till then, the old subscription will stay active.

If `messageId` changes, we want the previous subscription to old message to be removed

You can see the issue in action here: https://codesandbox.io/s/youthful-newton-k351z

I've replaced `willFocus` and `willBlur` events for `focus` and `blur` events on `window`, but it's pretty much the same. You can click on a button to change the ID, but the subscription for the old ID stays active until you unfocus the demo pane and refocus it.

Reply:
I see. I think the misunderstanding is related to what I was expecting useFocusEffect to do. I wasn't expecting it to unsub/resub if the id changes, but only to fire if a screen focus/unfocus (ie not unsub/resub if params.messageId is updated).

What about using `useFocusEffect(callback,[messageId])` instead?

Reply:
Also, how do you handle if unsub/resub is not so fast? Currently it will unsub/resub on every re-render, so it works fine (even when unsub/resub is not really needed because sub params did not change) but it starts to be a problem when subscribing gets more expensive (not sure but I'm thinking of things like starting a apollo watch query, which will try to hit the cache for data, eventually the network depending on your case)

Reply:
The problem with accepting a dep array is that the ESLint plugin doesn't recognize it as a dep array. In my usage of hooks, the ESLint plugin have been a necessity to make sure my deps are correct.

The React team recommends to pass a function wrapped with `useCallback` for this reason instead of accepting a dep array.

Reply:
I see thanks, didn't know that useCallback workaround pattern, will keep this in mind ;)

Reply:
hi @satya164 

I've posted a twitter thread about the stale closure problem where I illustrate with the issue above. If you have any insight to share I'd like to hear.
https://twitter.com/sebastienlorber/status/1178328607376232449

If we craft the hook api to always execute latest callback, I don't understand why we should care about the eslint rule, because the stale closure does not exist anymore so there's not really any trap, we are just expressing what variables should trigger a re-execution

Reply:
> If we craft the hook api to always execute latest callback, I don't understand why we should care about the eslint rule

The point of using `useCallback` instead is to work with the eslint rule. Sure our API could chose to not care about the ESLint rule, but the rule helps a lot of people like me ensure that we're not passing wrong deps. There's no good reason to ignore it.